packages/firestore/src/local/indexeddb_schema.ts

```diff
@@ -451,15 +474,87 @@ export class DbTargetGlobal {
      * until the backend has caught up to this snapshot version again. This
      * prevents our cache from ever going backwards in time.
      */
-    public lastRemoteSnapshotVersion: DbTimestamp
+    public lastRemoteSnapshotVersion: DbTimestamp,
+    /**
+     * The number of targets persisted.
+     */
+    public targetCount: number
   ) {}
 }
 
+function createQueryCache(db: IDBDatabase): void {
+  const targetDocumentsStore = db.createObjectStore(DbTargetDocument.store, {
+    keyPath: DbTargetDocument.keyPath as KeyPath
+  });
+  targetDocumentsStore.createIndex(
+    DbTargetDocument.documentTargetsIndex,
+    DbTargetDocument.documentTargetsKeyPath,
+    { unique: true }
+  );
+
+  const targetStore = db.createObjectStore(DbTarget.store, {
+    keyPath: DbTarget.keyPath
+  });
+
+  // NOTE: This is unique only because the TargetId is the suffix.
+  targetStore.createIndex(
+    DbTarget.queryTargetsIndexName,
+    DbTarget.queryTargetsKeyPath,
+    { unique: true }
+  );
+  db.createObjectStore(DbTargetGlobal.store);
+}
+
 /**
- * The list of all IndexedDB stored used by the SDK. This is used when creating
- * transactions so that access across all stores is done atomically.
+ * Counts the number of targets persisted and adds that value to the target
+ * global singleton.
+ *
+ * @param {IDBTransaction} txn The version upgrade transaction for indexeddb
  */
-export const ALL_STORES = [
+function saveTargetCount(txn: SimpleDbTransaction): PersistencePromise<void> {
+  const globalStore = txn.store<DbTargetGlobalKey, DbTargetGlobal>(
+    DbTargetGlobal.store
+  );
+  const targetStore = txn.store<DbTargetKey, DbTarget>(DbTarget.store);
+  return targetStore.count().next(count => {
+    return globalStore.get(DbTargetGlobal.key).next(metadata => {
+      metadata.targetCount = count;
+      return globalStore.put(DbTargetGlobal.key, metadata);
+    });
+  });
+}
+
+/**
+ * Ensures that the target global singleton row exists by adding it if it's
+ * missing.
+ *
+ * @param {IDBTransaction} txn The version upgrade transaction for indexeddb
+ */
+function ensureTargetGlobal(
```

Comment:
`ensureTargetGlobalEntryExists` ? The current name raises more questions than it answers :)

Reply:
Changed to `ensureTargetGlobalExists`. The comment should also clarify any questions raised.